logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -0,0 +1,145 @@
+# encoding: utf-8
+
+module LogStash
+  class PipelineState
+    attr_reader :pipeline_id, :pipeline
+
+    def initialize(pipeline_id, pipeline)
+      @pipeline_id = pipeline_id
+      @pipeline = pipeline
+      @reloading = Concurrent::AtomicBoolean.new(false)
+    end
+
+    def terminated?
+      # a reloading pipeline is never considered terminated
+      @reloading.false? && @pipeline.finished_execution?
+    end
+
+    def set_reloading(is_reloading)
+      @reloading.value = is_reloading
+    end
+
+    def set_pipeline(pipeline)
+      @pipeline = pipeline
+    end
+  end
+
+  class PipelinesRegistry
+    attr_reader :states
+    include LogStash::Util::Loggable
+
+    def initialize
+      # we leverage the semantic of the Java ConcurrentHashMap for the
+      # compute() method which is atomic; calling compute() concurrently
+      # will block until the other compute finishes so no mutex is necessary
+      # for synchronizing compute calls
+      @states = java.util.concurrent.ConcurrentHashMap.new
+    end
+
+    def create_pipeline(pipeline_id, pipeline, &create_block)
+      success = false
+
+      @states.compute(pipeline_id) do |_, state|
+        if state
+          if state.terminated?
+            success = yield
+            state.set_pipeline(pipeline)
+          else
+            logger.error("Attempted to create a pipeline that already exists", :pipeline_id => pipeline_id)
+          end
+          state
+        else
+          success = yield
+          success ? PipelineState.new(pipeline_id, pipeline) : nil
+        end
+      end
+
+      success
+    end
+
+    def terminate_pipeline(pipeline_id, &stop_block)
```

Comment:
It would be helpful to either converge on a single yield argument/return-value pattern for these three methods, or to at least document what [is yielded](https://www.rubydoc.info/gems/yard/file/docs/Tags.md#yieldparam), what [the block should return](https://www.rubydoc.info/gems/yard/file/docs/Tags.md#yieldreturn), and what the effect of certain return values is (if any).

As of this writing:
 - `create_pipeline` yields to the supplied block with no arguments, and expects the block to return boolean value indicating success
 - `terminate_pipeline` yields a `Pipeline` to the block, and ignores the block's return value
 - `reload_pipeline` yields the a `PipelineState`, expecting the block to return a two-tuple with a boolean value indicating success and a `Pipeline`.

Reply:
Yes, I agree, but on the other hand, this is not really a *reusable* API, these methods we're crafter for the specific need they serve in the context of the actions. I will add documentation per your suggestion, good idea.

Reply:
BTW I am not a super fan of this block passing strategy here, I chose that manly to minimize changes in the actual pipeline actions. I think I'd prefer to keep it that way for now but will add a note to see if we could clean that up which would probably include also refactoring the actions themselves.